token/program-2022/src/extension/confidential_mint_burn/instruction.rs

```diff
@@ -0,0 +1,582 @@
+    /// instruction has already executed for the mint.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   0. `[writable]` The SPL Token mint.
+    ///
+    /// Data expected by this instruction:
+    ///   `InitializeMintData`
+    InitializeMint,
+    /// Updates mint-authority for confidential-mint-burn mint.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   * Single authority
+    ///   0. `[writable]` The SPL Token mint.
+    ///   1. `[signer]` Confidential mint authority.
+    ///
+    ///   * Multisignature authority
+    ///   0. `[writable]` The SPL Token mint.
+    ///   1. `[]` The multisig authority account owner.
+    ///   2.. `[signer]` Required M signer accounts for the SPL Token Multisig
+    ///
+    /// Data expected by this instruction:
+    ///   `UpdateAuthorityData`
+    UpdateAuthority,
+    /// Rotates the ElGamal pubkey used to encrypt confidential supply
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   * Single authority
+    ///   0. `[writable]` The SPL Token mint.
+    ///   1. `[signer]` Confidential mint authority.
+    ///
+    ///   * Multisignature authority
+    ///   0. `[writable]` The SPL Token mint.
+    ///   1. `[]` The multisig authority account owner.
+    ///   2.. `[signer]` Required M signer accounts for the SPL Token Multisig
+    ///
+    /// Data expected by this instruction:
+    ///   `RotateSupplyElGamalPubkeyData`
+    RotateSupplyElGamalPubkey,
+    /// Updates the decryptable supply of the mint
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   * Single authority
+    ///   0. `[writable]` The SPL Token mint.
+    ///   1. `[]` Instructions sysvar if `CiphertextCiphertextEquality` is
+    ///      included in the same transaction or context state account if
+    ///      `CiphertextCiphertextEquality` is pre-verified into a context state
+    ///      account.
+    ///   2. `[signer]` Confidential mint authority.
+    ///
+    ///   * Multisignature authority
+    ///   0. `[writable]` The SPL Token mint.
+    ///   1. `[]` Instructions sysvar if `CiphertextCiphertextEquality` is
+    ///      included in the same transaction or context state account if
+    ///      `CiphertextCiphertextEquality` is pre-verified into a context state
+    ///      account.
+    ///   2. `[]` The multisig authority account owner.
+    ///   3.. `[signer]` Required M signer accounts for the SPL Token Multisig
+    ///
+    /// Data expected by this instruction:
+    ///   `UpdateDecryptableSupplyData`
+    UpdateDecryptableSupply,
+    /// Mints tokens to confidential balance
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   * Single authority
+    ///   0. `[writable]` The SPL Token account.
+    ///   1. `[]` The SPL Token mint. `[writable]` if the mint has a non-zero
+    ///      supply elgamal-pubkey
+    ///   2. `[]` The context state account containing the pre-verified
```

Comment:
I think we should add the sysvar account here? Let's also maybe use [this](https://github.com/solana-labs/solana-program-library/blob/master/token/program-2022/src/extension/confidential_transfer/instruction.rs#L269-L277) style of specifying the instructions to be consistent. Let's also make the proof accounts `Optional` in case transaction size limits are lifted in the future and we can include all the proofs in a single tx.